mlir/lib/Conversion/ConvertToLLVM/ConvertToLLVMPass.cpp

```diff
@@ -104,20 +99,33 @@ class ConvertToLLVMPass
         auto *iface = dyn_cast<ConvertToLLVMPatternInterface>(dialect);
         if (!iface)
           continue;
-        iface->populateConvertToLLVMConversionPatterns(*target, *typeConverter,
-                                                       tempPatterns);
+        interfaces->push_back(iface);
       }
     }
 
-    this->patterns =
-        std::make_unique<FrozenRewritePatternSet>(std::move(tempPatterns));
-    this->target = target;
-    this->typeConverter = typeConverter;
+    this->interfaces = interfaces;
     return success();
   }
 
   void runOnOperation() final {
-    if (failed(applyPartialConversion(getOperation(), *target, *patterns)))
+    MLIRContext *context = &getContext();
+    RewritePatternSet patterns(context);
+    ConversionTarget target(*context);
+    target.addLegalDialect<LLVM::LLVMDialect>();
+    LLVMTypeConverter typeConverter(context);
```

Comment:
I'm concerned about the cost of all this at the moment :(

Reply:
I can profile it and check the impact.

However, let's say the cost of that initialization is `C`. In the previous version `C` is paid only once.
In this version the overall cost is `C` * `threads`, so the overall computation time does increment.
However, on a thread per thread basis the cost remains relatively the same.

So, I would argue, that unless the effect of the change is severe, the extra overhead might be justified for the feature.

Reply:
Test results:
LLVM Release with assertions off 
cpu: Intel(R) Xeon(R) Gold 6132 CPU @ 2.60GHz
Test: run 100 times `mlir-opt --pass-pipeline="builtin.module(convert-to-llvm)"  mlir/test/Conversion/FuncToLLVM/func-to-llvm.mlir --mlir-timing` 

convert-to-llvm time:
```
Pre change: 0.00210028567 s
Post change: 0.0019316432 s
Difference: 0.00016864247 s
Slowdown(post/pre): 1.08730518658932
```

Total time:
```
Pre change: 0.00816961453 s
Post change: 0.00827812259 s
Difference: 0.00010850806 s
Slowdown(post/pre): 1.01328190694941
```

Globally the impact seems to be minimal (less than 2%), per pass it seems to increase around 10%. However, the per pass time doesn't include initialization.

Reply:
> However, let's say the cost of that initialization is C. In the previous version C is paid only once. In this version the overall cost is C * threads, so the overall computation time does increase. However, on a thread per thread basis the cost might not be as impactful.

I believe the cost is `C * thread * N` where N is the number of invocation of the pass-pipeline: if I setup a compilation server my pass-pipeline could be initialized once and used to compiled many modules.

Reply:
I also thought that could be the case. However, it appears that `initialize` is called every time. I added the pass multiple times in a pipeline and added a print in the initialize method of the method,  and the method is called every time. Seems like something to fix in the `PassManager`.

But even then, let's assume that the cost of one vs the other is `thread * N` times greater. If `convert-to-llvm` was a pass like `cse` or `canonicalize` I would 100% agree with you, however, this is a pass that likely won't be invoked that many times.

Reply:
I can think 2 solutions to this issue with little to no performance impact. 
Create a new `convert-to-llvm` like pass, where we say the performance cost doesn't matter. 
Make `ConversionTarget`, `TypeConverter` and `RewritePatternSet` somewhat copyable.